packages/stylist-macros/src/inline/css_ident.rs

```diff
@@ -0,0 +1,119 @@
+use proc_macro2::{Punct, Spacing, TokenStream};
+use quote::ToTokens;
+use std::fmt::{Display, Formatter};
+use syn::{
+    ext::IdentExt,
+    parse::{Parse, ParseBuffer, Result as ParseResult},
+    spanned::Spanned,
+    token, Ident, LitStr,
+};
+
+syn::custom_punctuation!(DoubleSub, --);
+
+#[derive(Debug, Clone)]
+pub enum IdentPart {
+    Dash(Punct),
+    Ident(Ident),
+}
+
+#[derive(Debug, Clone)]
+pub struct CssIdent {
+    parts: Vec<IdentPart>,
+}
+
+impl IdentPart {
+    pub fn peek(lookahead: &ParseBuffer, accept_dash: bool, accept_ident: bool) -> bool {
+        let peek_dash = accept_dash && (lookahead.peek(token::Sub) || lookahead.peek(DoubleSub));
+        let peek_ident = accept_ident && lookahead.peek(Ident::peek_any);
+        peek_dash || peek_ident
+    }
+}
+
+impl Display for CssIdent {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        let name = self.to_name_string();
+        f.write_str(&name)
+    }
+}
+
+impl CssIdent {
+    pub fn peek(lookahead: &ParseBuffer) -> bool {
+        if lookahead.peek(token::Sub) {
+            // A single dash is not an identifier
+            lookahead.peek2(token::Sub) || lookahead.peek2(Ident::peek_any)
+        } else {
+            IdentPart::peek(lookahead, true, true)
+        }
+    }
+
+    pub fn to_name_string(&self) -> String {
```

Comment:
That was intentional: I think a Display implementation should be what appears in errors if verification of attribute names, et al is implementation in the future and differs, at least conceptually, from the one that returns the contents of the string literal used in quote.

Reply:
In that case, you can still use `format!("{}");`. If clippy lints the generated code, you can `#[allow(clippy::useless_format)]`.

See: https://rust-lang.github.io/rust-clippy/stable/index.html

Reply:
I see it like this:
- `Debug`: what we want to see if we debug the macro
- `Display`: what the user sees if the `CssIdent` shows up in an error message
- `to_name_string`: what gets written into the output css

I'm not sure either pair of these three should always result in the same string.

Reply:
> Display: what the user sees if the CssIdent shows up in an error message (currently exactly what to_name_string returns. I now see what you mean, but I don't want to commit to it staying that way)

I cannot say you are wrong, but with this understanding we cannot explain why interpolated values use `Display` value when printed into the stylesheet.

What you are looking for here is probably a Python's `repr(x)` equivalent which returns a representation of value when invoked. Which in rust, it is usually `{:?}` (Debug).

Reply:
> What you are looking for here is probably a Python's `repr(x)` equivalent which returns a representation of value when invoked. Which in rust, it is usually `{:?}` (Debug). Not strictly equivalent, but usually so.

What I try to say is connected to `trycase` (which I have removed for now for simplicity). Consider:
```
background_color: black;
```
Currently not supported, cause I removed the case transformation, but I did already get a question regarding that on discord. In any case, as implemented in this pr:
```
Debug: "CssIdent { parts: [IdentPart::Ident("background_color")] }"
Display: "background_color"
to_name_string: "background_color"
```
with case transformation added
```
Debug: "CssIdent { parts: [IdentPart::Ident("background_color")] }"
Display: "background_color" <-- this should show in most error messages, as it's what the user wrote?
to_name_string: "background-color" <-- this gets written into the css
```

> with this understanding we cannot explain why interpolated values use Display value when printed into a stylesheet.

I'm not sure I understand what interpolated values have to do with `CssIdent`, but in either case I do think that at some point switching to a custom trait instead of reusing `Display` could be beneficial.

Reply:
Hmm, I see.

My rationale is always that we don't have to consider a feature until we actually lands it so our code base is optimised for what we have at the moment not something we don't have at the moment.

But I guess `to_output_string` would be good enough.

So Css* variant is more based on what's in the code, and `Output*` is more based on what's going to be in `stylist::ast`.

Reply:
> So Css* variant is more based on what's in the code, and `Output*` is more based on what's going to be in `stylist::ast`.

Correct, that's what it's supposed to mean.